src/hotspot/share/opto/type.cpp

```diff
@@ -1745,218 +1748,120 @@ const TypeInteger* TypeInteger::minus_1(BasicType bt) {
-    return Type::BOTTOM;
-  default:                      // All else is a mistake
-    typerr(t);
-  case Top:                     // No change
-    return this;
-  case Int:                     // Int vs Int?
-    break;
-  }
-
-  // Expand covered set
-  const TypeInt *r = t->is_int();
-  return make( MIN2(_lo,r->_lo), MAX2(_hi,r->_hi), MAX2(_widen,r->_widen) );
+const TypeInt* TypeInt::make(jint lo, jint hi, int widen) {
+  assert(lo <= hi, "must be legal bounds");
+  return try_make(TypeIntPrototype<jint, juint>{{lo, hi}, {0, max_juint}, {0, 0}}, widen)->is_int();
 }
 
-//------------------------------xdual------------------------------------------
-// Dual: reverse hi & lo; flip widen
-const Type *TypeInt::xdual() const {
-  int w = normalize_int_widen(_hi,_lo, WidenMax-_widen);
-  return new TypeInt(_hi,_lo,w);
-}
-
-//------------------------------widen------------------------------------------
-// Only happens for optimistic top-down optimizations.
-const Type *TypeInt::widen( const Type *old, const Type* limit ) const {
-  // Coming from TOP or such; no widening
-  if( old->base() != Int ) return this;
-  const TypeInt *ot = old->is_int();
-
-  // If new guy is equal to old guy, no widening
-  if( _lo == ot->_lo && _hi == ot->_hi )
-    return old;
-
-  // If new guy contains old, then we widened
-  if( _lo <= ot->_lo && _hi >= ot->_hi ) {
-    // New contains old
-    // If new guy is already wider than old, no widening
-    if( _widen > ot->_widen ) return this;
-    // If old guy was a constant, do not bother
-    if (ot->_lo == ot->_hi)  return this;
-    // Now widen new guy.
-    // Check for widening too far
-    if (_widen == WidenMax) {
-      int max = max_jint;
-      int min = min_jint;
-      if (limit->isa_int()) {
-        max = limit->is_int()->_hi;
-        min = limit->is_int()->_lo;
-      }
-      if (min < _lo && _hi < max) {
-        // If neither endpoint is extremal yet, push out the endpoint
-        // which is closer to its respective limit.
-        if (_lo >= 0 ||                 // easy common case
-            ((juint)_lo - min) >= ((juint)max - _hi)) {
-          // Try to widen to an unsigned range type of 31 bits:
-          return make(_lo, max, WidenMax);
-        } else {
-          return make(min, _hi, WidenMax);
-        }
-      }
-      return TypeInt::INT;
-    }
-    // Returned widened new guy
-    return make(_lo,_hi,_widen+1);
-  }
-
-  // If old guy contains new, then we probably widened too far & dropped to
-  // bottom.  Return the wider fellow.
-  if ( ot->_lo <= _lo && ot->_hi >= _hi )
-    return old;
+const Type* TypeInt::try_make(const TypeIntPrototype<jint, juint>& t, int widen) {
+  return try_make(t, widen, false);
+}
 
-  //fatal("Integer value range is not subset");
-  //return this;
-  return TypeInt::INT;
+bool TypeInt::contains(jint i) const {
+  juint u = i;
+  return i >= _lo && i <= _hi && u >= _ulo && u <= _uhi && _bits.is_satisfied_by(u);
```

Comment:
```suggestion
  return _lo  <= i && i <= _hi  &&
         _ulo <= u && u <= _uhi &&
         _bits.is_satisfied_by(u);
```
Optional, for readability.